### PR TITLE
Upgrade toolchain to LLVM17 & Support C++23

### LINKED ISSUE #7193: Clang 17 & C++23 Compatibility
## Enhancement

Clang 16.0.0 was officially released 2 weeks ago (March, 2023).

https://github.com/llvm/llvm-project/releases/tag/llvmorg-16.0.0

### LINKED ISSUE #8898: Upgrade C++ standard to C++23
## Enhancement

https://en.cppreference.com/w/cpp/23

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: close #8898, close #7193

Problem Summary: upgrade the toolchain to llvm 17.0.6 for better code diagnostic

### What is changed and how it works?

tested the performance of TPC-H and TPC-DS under amd64

#### TPC-H
|       | Clang13 | Clang17 | diff  | diff by percentage |
|-------|---------|---------|-------|--------------------|
| Total | 64.277  | 59.913  | 4.364 | 6.79               |
| Q19   | 2.039   | 1.42    | 0.619 | 30.36              |
| Q12   | 2       | 1.536   | 0.465 | 23.23              |
| Q9    | 12.333  | 11.872  | 0.461 | 3.74               |
| Q6    | 0.959   | 0.641   | 0.318 | 33.15              |
| Q14   | 1.059   | 0.742   | 0.317 | 29.96              |
| Q15   | 1.067   | 0.759   | 0.308 | 28.89              |
| Q21   | 6.394   | 6.127   | 0.266 | 4.17               |
| Q18   | 6.59    | 6.353   | 0.237 | 3.6                |
| Q5    | 4.06    | 3.863   | 0.197 | 4.85               |
| Q10   | 2.425   | 2.247   | 0.178 | 7.35               |
| Q20   | 1.414   | 1.241   | 0.173 | 12.24              |
| Q7    | 2.473   | 2.319   | 0.154 | 6.24               |
| Q8    | 3.379   | 3.237   | 0.141 | 4.18               |
| Q4    | 1.804   | 1.673   | 0.13  | 7.23               |
| Q1    | 2.862   | 2.775   | 0.087 | 3.03               |
| Q17   | 5.663   | 5.582   | 0.081 | 1.43               |
| Q13   | 3.306   | 3.244   | 0.062 | 1.86               |
| Q2    | 0.903   | 0.843   | 0.06  | 6.62               |
| Q3    | 2.164   | 2.107   | 0.057 | 2.64               |
| Q22   | 0.41    | 0.387   | 0.023 | 5.54               |
| Q11   | 0.404   | 0.389   | 0.015 | 3.74               |
| Q16   | 0.569   | 0.556   | 0.014 | 2.39               |

#### TPC-DS
|            | Clang13 | Clang17 | diff   | diff by percentage                 |
|------------|---------|---------|--------|------------------------------------|
| Total      | 630.154 | 668.895 | 38.741 |                               6.15 |
| QUERY_95   |  506.15 | 540.400 |  34.25 | 6.77*<br>(check the details below) |
| QUERY_72   |   5.032 |   5.767 |  0.734 |                              14.59 |
| QUERY_53   |   3.142 |   3.592 |   0.45 |                              14.32 |
| QUERY_63   |   2.748 |   3.085 |  0.338 |                              12.28 |
| QUERY_51   |  46.575 |    46.9 |  0.325 |                                0.7 |
| QUERY_65   |   3.878 |   4.132 |  0.255 |                               6.58 |
| QUERY_74   |   2.765 |    3.01 |  0.245 |                               8.86 |
| QUERY_41   |    7.77 |    7.97 |    0.2 |                               2.57 |
| QUERY_96   |   0.548 |   0.382 | -0.166 |                             -30.35 |
| QUERY_64   |   2.625 |    2.79 |  0.165 |                               6.29 |
| QUERY_3    |   0.367 |   0.502 |  0.135 |                              36.85 |
| QUERY_75   |   2.402 |   2.537 |  0.134 |                               5.58 |
| QUERY_88   |   2.548 |    2.65 |  0.102 |                               4.02 |
| QUERY_47   |   2.402 |     2.5 |  0.098 |                               4.06 |
| QUERY_79   |   1.365 |   1.457 |  0.092 |                               6.72 |
| QUERY_28   |   1.628 |   1.715 |  0.088 |                               5.38 |
| QUERY_46   |   1.165 |    1.25 |  0.085 |                                7.3 |
| QUERY_68   |   0.971 |    1.05 |  0.079 |                               8.13 |
| QUERY_25   |   1.185 |    1.25 |  0.065 |                               5.49 |
| QUERY_7    |   0.945 |   1.006 |  0.061 |                               6.48 |
| QUERY_17   |   1.448 |   1.502 |  0.055 |                                3.8 |
| QUERY_31   |   0.604 |   0.657 |  0.053 |                               8.74 |
| QUERY_1    |   1.802 |    1.85 |  0.048 |                               2.64 |
| QUERY_13   |    1.82 |   1.865 |  0.045 |                               2.47 |
| QUERY_50   |   0.471 |   0.514 |  0.043 |                               9.18 |
| QUERY_48   |   1.908 |    1.95 |  0.042 |                               2.23 |
| QUERY_29   |   1.222 |   1.265 |  0.042 |                               3.48 |
| QUERY_89   |   1.175 |   1.213 |  0.038 |                               3.26 |
| QUERY_24_0 |   0.616 |   0.654 |  0.038 |                               6.13 |
| QUERY_24_1 |    0.61 |   0.647 |  0.037 |                               6.02 |
| QUERY_26   |   0.413 |   0.449 |  0.036 |                               8.73 |
| QUERY_10   |   0.473 |   0.506 |  0.034 |                               7.12 |
| QUERY_49   |    0.82 |   0.853 |  0.033 |                               3.99 |
| QUERY_94   |   0.287 |    0.32 |  0.033 |                              11.44 |
| QUERY_43   |   0.593 |   0.624 |  0.031 |                               5.22 |
| QUERY_69   |   0.408 |   0.438 |   0.03 |                               7.27 |
| QUERY_60   |    0.61 |    0.64 |   0.03 |                               4.91 |
| QUERY_66   |   0.632 |   0.658 |  0.027 |                                4.2 |
| QUERY_35   |   0.731 |   0.756 |  0.025 |                               3.42 |
| QUERY_15   |   0.188 |   0.214 |  0.025 |                               13.4 |
| QUERY_71   |   0.597 |   0.622 |  0.024 |                               4.06 |
| QUERY_73   |   0.505 |   0.525 |  0.021 |                               4.07 |
| QUERY_81   |   0.235 |   0.256 |   0.02 |                               8.57 |
| QUERY_34   |   0.783 |   0.802 |  0.019 |                               2.38 |
| QUERY_99   |   0.863 |   0.881 |  0.018 |                               2.08 |
| QUERY_61   |     0.9 |   0.918 |  0.018 |                               2.02 |
| QUERY_21   |    0.34 |   0.323 | -0.017 |                              -4.95 |
| QUERY_91   |   0.176 |    0.16 | -0.016 |                              -9.04 |
| QUERY_84   |   0.133 |   0.147 |  0.015 |                              10.98 |
| QUERY_56   |   0.552 |   0.565 |  0.014 |                               2.45 |
| QUERY_93   |   0.681 |   0.667 | -0.014 |                              -2.03 |
| QUERY_76   |   0.343 |   0.358 |  0.014 |                                4.1 |
| QUERY_19   |   0.518 |   0.531 |  0.013 |                               2.54 |
| QUERY_33   |   0.576 |   0.588 |  0.012 |                               2.06 |
| QUERY_45   |   0.202 |   0.214 |  0.012 |                                6.1 |
| QUERY_62   |   0.446 |   0.459 |  0.012 |                               2.77 |
| QUERY_6    |    0.35 |   0.361 |  0.011 |                               3.19 |
| QUERY_16   |   0.682 |   0.693 |  0.011 |                               1.67 |
| QUERY_98   |   0.344 |   0.353 |  0.009 |                               2.51 |
| QUERY_54   |   0.533 |   0.542 |  0.009 |                               1.64 |
| QUERY_30   |   0.252 |    0.26 |  0.008 |                               3.28 |
| QUERY_44   |   0.275 |   0.283 |  0.008 |                               2.83 |
| QUERY_37   |   0.207 |   0.215 |  0.007 |                               3.61 |
| QUERY_90   |   0.148 |   0.142 | -0.006 |                              -4.18 |
| QUERY_92   |   0.139 |   0.145 |  0.006 |                               4.35 |
| QUERY_40   |   0.426 |   0.432 |  0.006 |                                1.5 |
| QUERY_42   |   0.268 |   0.263 | -0.005 |                              -1.76 |
| QUERY_52   |   0.266 |   0.271 |  0.005 |                               1.86 |
| QUERY_57   |   1.018 |   1.022 |  0.005 |                               0.49 |
| QUERY_58   |     0.4 |   0.406 |  0.005 |                               1.36 |
| QUERY_83   |   0.302 |   0.306 |  0.004 |                               1.26 |
| QUERY_87   |   2.042 |   2.047 |  0.004 |                                0.2 |
| QUERY_12   |    0.12 |   0.123 |  0.004 |                               3.03 |
| QUERY_82   |   0.262 |   0.265 |  0.003 |                               1.34 |
| QUERY_55   |   0.245 |   0.248 |  0.003 |                               1.05 |
| QUERY_20   |    0.16 |   0.159 | -0.002 |                              -1.01 |
| QUERY_8    |   0.589 |    0.59 |  0.001 |                               0.12 |
| QUERY_32   |   0.233 |   0.232 | -0.001 |                              -0.35 |
| QUERY_9    |   0.001 |   0.001 | 0      |                             -35.18 |

##### TPC-DS query95
The difference in execution times of Round 2 and Round 3 is small. The main difference is in execution time for Round 1. In Round 1, the main difference is HashJoin_232, which is executed in TiDB.
|         | Clang13 | Clang17 |
|---------|---------|---------|
| Round 1 | 691.10  | 725.30  |
| Round 2 | 444.00  | 448.20  |
| Round 3 | 444.90  | 447.70  |

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Upgrade toolchain to LLVM17 & Support C++23
```

### CODE REVIEW

File: dbms/src/Storages/S3/FileCache.h [hunk 1]
```diff
@@ -227,8 +228,15 @@ class FileCache
 public:
 #endif
 
+    static FileCache & initializeInstance(
+        PathCapacityMetricsPtr capacity_metrics_,
+        const StorageRemoteCacheConfig & config_)
+    {
+        static FileCache instance(capacity_metrics_, config_);
+        return instance;
+    }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/S3/FileCache.h:237]: why does it need to be changed from a static uniqut_ptr to a static instance?

File: dbms/src/Storages/S3/FileCache.h [hunk 2]
```diff
@@ -201,17 +201,18 @@ class FileCache
...
     {
-        return global_file_cache_initialized.load(std::memory_order_acquire) ? global_file_cache_instance.get()
-                                                                             : nullptr;
+        return global_file_cache_initialized.load(std::memory_order_acquire)
+            ? &initializeInstance(nullptr, {}) // Dummy arguments, static variables are already initialized.
+            : nullptr;
     }
 
-    static void shutdown() { global_file_cache_instance = nullptr; }
+    static void shutdown() { instance()->~FileCache(); } // Call destructor to destroy the global instance.
```
Comment (CalvinNeo) [inline dbms/src/Storages/S3/FileCache.h:215]: Why explicitly destruct a global singleton and make it still accessible afterwards? If its lifetime is shorter, make it not accessible after the shutdown.
If some of its resources should be reclaimed before the programs shuts down, then reclaim it in `FileCache` explicitly.

File: dbms/src/Storages/S3/FileCache.h (hunk 1 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/S3/FileCache.h:237]: unique_ptr can't delete an incomplete type

File: dbms/src/Storages/S3/FileCache.h (hunk 2 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/S3/FileCache.h]: Add a new `GlobalFileCacheInstance` to resolve "unique_ptr can't delete an incomplete type"

File: dbms/src/Storages/S3/FileCache.h (hunk 1 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/S3/FileCache.h]: Reverted and add a new GlobalFileCacheInstance.

**Comment by zhangjinpeng87** [comment]
@Lloyd-Pottiger How about hold this PR until 8.1LTS released?

**Comment by Lloyd-Pottiger** [comment]
/hold

hold until 8.1LTS released

File: dbms/src/Storages/KVStore/tests/gtest_kvstore.cpp [hunk 3]
```diff
@@ -1542,7 +1544,9 @@ TEST_F(RegionKVStoreOldTest, RegionRange)
         catch (Exception & e)
         {
             const auto & res = e.message();
-            ASSERT_EQ(res, "void DB::RegionsRangeIndex::remove(const RegionRange &, RegionID): not found region_id=1");
+            std::regex msg_reg(
+                "void DB::RegionsRangeIndex::remove\\(const .*RegionRange &, .*RegionID\\): not found region_id=1");
```
Comment (JaySon-Huang) [inline dbms/src/Storages/KVStore/tests/gtest_kvstore.cpp:1548]:
```suggestion
                ".*: not found region_id=1");
```
Replacing the whole `__PRETTERY_FUNCTION__` by regex match is better

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/KVStore/tests/gtest_kvstore.cpp:1548]: Then we can remove `__PRETTERY_FUNCTION__` directly?

Comment (JaySon-Huang) [reply dbms/src/Storages/KVStore/tests/gtest_kvstore.cpp:1548]: "not found region_id=xxx" does not contain sufficient info about where the exception is thrown. If we catch and log the exception without its stacktrack, it is hard to find out where the error message come from.

While in this unit test, we just wan to ensure the left part contain the correct region_id without caring about the function name

File: dbms/src/IO/BaseFile/WritableFile.h [hunk 4]
```diff
@@ -14,6 +14,9 @@
 
 #pragma once
 
+#if __APPLE__ && __clang__
+#include <sys/_types/_off_t.h>
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/IO/BaseFile/WritableFile.h]: fixed

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Comment by wuhuizuo** [comment]
/hold

wait for more detail votes.

**Review by zanmato1984 — APPROVED** [review]

**Review by purelind — APPROVED** [review]
/lgtm

**Review by yudongusa — APPROVED** [review]
LGTM

**Pull request merged**